src/libopensc/card-entersafe.c

```diff
@@ -468,7 +496,7 @@ static int entersafe_select_fid(sc_card_t *card,
 	SC_TEST_RET(card->ctx, SC_LOG_DEBUG_NORMAL, r, "APDU transmit failed");
 
 	/* update cache */
-	if (file && file->type == SC_FILE_TYPE_DF) {
+	if (file->type == SC_FILE_TYPE_DF) {
```

Comment:
You need to either `sc_file_free(file)` in case of an error or you remove the local `file` by passing `file_out` to `iso_ops->select_file`. Currently there is a memory leak in your code.